peer.go

```diff
@@ -74,81 +83,76 @@ func (l *PeerList) newSibling() *PeerList {
 
 // Add adds a peer to the list if it does not exist, or returns any existing peer.
 func (l *PeerList) Add(hostPort string) *Peer {
-	l.RLock()
-
-	if p, ok := l.peersByHostPort[hostPort]; ok {
-		l.RUnlock()
-		return p
+	if ps, ok := l.exists(hostPort); ok {
+		return ps.Peer
 	}
-
-	l.RUnlock()
 	l.Lock()
 	defer l.Unlock()
 
 	if p, ok := l.peersByHostPort[hostPort]; ok {
-		return p
+		return p.Peer
 	}
 
 	p := l.parent.Add(hostPort)
-	atomic.AddUint32(&p.scCount, 1)
-	l.peersByHostPort[hostPort] = p
-	l.peers = append(l.peers, p)
+	ps := newPeerScore(p)
+	ps.score = l.scoreCalculator.GetScore(p)
+
+	l.peersByHostPort[hostPort] = ps
+	l.peerHeap.PushPeer(ps)
 	return p
 }
 
 // Get returns a peer from the peer list, or nil if none can be found.
-// Get will avoid peers in prevSelected unless there are no other peers left.
 func (l *PeerList) Get(prevSelected map[string]struct{}) (*Peer, error) {
-	l.RLock()
-
-	if len(l.peers) == 0 {
-		l.RUnlock()
+	l.Lock()
+	if l.peerHeap.Len() == 0 {
+		l.Unlock()
 		return nil, ErrNoPeers
 	}
 
 	// Select a peer, avoiding previously selected peers. If all peers have been previously
 	// selected, then it's OK to repick them.
-	peer := l.choosePeer(l.peers, prevSelected)
+	peer := l.choosePeer(prevSelected)
 	if peer == nil {
-		peer = l.choosePeer(l.peers, nil)
+		peer = l.choosePeer(nil)
 	}
-	l.RUnlock()
-
+	l.Unlock()
 	return peer, nil
 }
 
-func (l *PeerList) choosePeer(peers []*Peer, prevSelected map[string]struct{}) *Peer {
-	var maxScore uint64
-	var bestPeer *Peer
+func (l *PeerList) choosePeer(prevSelected map[string]struct{}) *Peer {
+	var psPopList []*peerScore
+	var ps *peerScore
 
-	// pick peer with highest score that is not in prevSelected.
-	for _, peer := range peers {
-		if _, ok := prevSelected[peer.HostPort()]; ok {
-			continue
+	size := l.peerHeap.Len()
+	for i := 0; i < size; i++ {
+		ps = l.peerHeap.PopPeer()
+		if _, ok := prevSelected[ps.Peer.HostPort()]; !ok {
+			break
 		}
+		psPopList = append(psPopList, ps)
+	}
 
-		score := l.scoreCalculator.GetScore(peer)
-		if score > maxScore {
-			maxScore = score
-			bestPeer = peer
-		}
+	for _, p := range psPopList {
+		heap.Push(l.peerHeap, p)
 	}
 
-	if bestPeer != nil {
-		atomic.AddUint64(&bestPeer.chosenCount, 1)
+	if ps == nil {
+		return nil
 	}
-	return bestPeer
+	ps.Lock()
+	ps.score++
```

Comment:
We should probably break this logic up -- seems like this should belong in the score calculator